zerver/tests/test_subs.py

```diff
@@ -2198,6 +2204,86 @@ def test_unarchive_stream(self) -> None:
             },
         )
 
+    def test_permissions_and_archiving_behavior(self) -> None:
+        """
+        Test permissions for archiving and unarchiving streams, and ensure users without
+        the necessary permissions cannot archive or unarchive a stream.
+        """
+        desdemona = self.example_user("desdemona")
+
+        iago = self.example_user("iago")
+
+        self.login_user(desdemona)
+        stream = self.make_stream("test_stream", invite_only=False)
+        self.subscribe(iago, stream.name)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_success(result)
+        stream.refresh_from_db()
+        self.assertTrue(stream.deactivated)
+
+        result = self.client_patch(
+            f"/json/streams/{stream.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_success(result)
+
+        cordellia = self.example_user("cordelia")
+        self.login_user(cordellia)
+
+        stream1 = self.make_stream("test_stream_1", invite_only=False)
+        self.subscribe(iago, stream1.name)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+        stream1.refresh_from_db()
+        self.assertTrue(stream.deactivated)
+
+        result = self.client_patch(
```

Comment:
This checks unarchiving a stream which is not archived. A better way to perform a task which is not being checked in tests, is to directly use `do_..` function, `do_deactivate_stream` in this case, so that permission checks can be bypassed.

Reply:
And it also makes it easy to review the code as it is easy to check only the `self.client_` calls to see which cases are being tested. So, I think would be good to just use `do_deactivate_stream` directly in this test in all places where you want to archive the stream instead of using `client_delete`.